x/leverage/keeper/exchangerate.go

```diff
@@ -0,0 +1,118 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/umee-network/umee/x/leverage/types"
+)
+
+// ExchangeTokens converts an sdk.Coin containing base assets to its value in uTokens
+func (k Keeper) ExchangeTokens(ctx sdk.Context, tokens sdk.Coin) (sdk.Coin, error) {
+	uTokenDenom := k.FromTokenToUTokenDenom(ctx, tokens.Denom)
+	if uTokenDenom == "" {
+		return sdk.Coin{}, sdkerrors.Wrap(types.ErrInvalidAsset, tokens.Denom)
+	}
+	exchangeRate, err := k.GetExchangeRate(ctx, tokens.Denom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	uTokenAmount := tokens.Amount.ToDec().Quo(exchangeRate).TruncateInt()
+	return sdk.NewCoin(uTokenDenom, uTokenAmount), nil
+}
+
+// ExchangeUTokens converts an sdk.Coin containing uTokens to its value in base assets
+func (k Keeper) ExchangeUTokens(ctx sdk.Context, utokens sdk.Coin) (sdk.Coin, error) {
+	tokenDenom := k.FromUTokenToTokenDenom(ctx, utokens.Denom)
+	if tokenDenom == "" {
+		return sdk.Coin{}, sdkerrors.Wrap(types.ErrInvalidAsset, utokens.Denom)
+	}
+	exchangeRate, err := k.GetExchangeRate(ctx, tokenDenom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	tokenAmount := utokens.Amount.ToDec().Mul(exchangeRate).TruncateInt()
+	return sdk.NewCoin(tokenDenom, tokenAmount), nil
+}
+
+// GetExchangeRate gets the token:uTokenexchange rate for a given base token denom
+func (k Keeper) GetExchangeRate(ctx sdk.Context, denom string) (sdk.Dec, error) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.CreateExchangeRateKey(denom)
+	amount := sdk.ZeroDec()
+	bz := store.Get(key)
+	if bz == nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrInvalidAsset, denom)
+	}
+	err := amount.Unmarshal(bz)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
Why does this return an error and panic?

Reply:
Fixed, along with the rest of the panic(err) instances in error-returning functions in `x/leverage`